File: python/packages/core/src/varar_core/failure.py
"""failure.py — port of typescript/packages/core/src/failure.ts.

Converts a thrown step error into the structured ExampleFailure payload.
Shared by every producer so failures are byte-identical.
"""
from __future__ import annotations

import re
from typing import Any

from varar_core.cell_diff import is_cell_mismatch_error
from varar_core.result import CellFailure, ExampleFailure


def _failing_line(stack: str, oath_path: str) -> int | None:
    """Recover the 1-based failing line from a ``<oathPath>:line:col`` frame."""
    escaped = re.escape(oath_path)
    m = re.search(rf"{escaped}:(\d+):\d+", stack)
    return int(m.group(1)) if m else None


def to_failure(
    error: Any,
    oath_path: str,
    fallback_line: int,
) -> ExampleFailure:
    """Convert a thrown step error to an ExampleFailure.

    Checks for ``CellMismatchError`` first to
    populate structured ``cells``/``doc`` payloads; falls back to a plain
    message + stack for any other exception.

    ``error.stack`` (if present as a string attribute) is used verbatim — this
    matches the TypeScript port where execute.ts injects a ``<oathPath>:line:col``
    frame that ``_failing_line`` then extracts.
    """
    # Prefer a manually-set .stack attribute (mirrors TS Error.stack behaviour).
    raw_stack = getattr(error, "stack", None)
    stack: str = raw_stack if isinstance(raw_stack, str) else str(error)
    message: str = str(error)

    cells: tuple[CellFailure, ...] | None = None
    if is_cell_mismatch_error(error):
        failing = tuple(
            CellFailure(
                from_=c.span.start_offset,
                to=c.span.end_offset,
                actual=c.actual,
            )
            for c in error.cells
            if not c.ok
        )
        if failing:
            cells = failing

    line = _failing_line(stack, oath_path) if stack else None

    return ExampleFailure(
        line=line if line is not None else fallback_line,
        message=message,
        stack=stack,
        cells=cells,
    )

File: typescript/packages/core/src/failure.ts
import { isCellMismatchError } from './cell-diff.ts'
import { readFailureAnchor } from './failure-anchor.ts'
import type { CellFailure, ExampleResult } from './result.ts'

// Recover the 1-based failing line from the `<oathPath>:line:col` frame
// executePlan injects (see execute.ts augmentStack). Internal — not exported
// from the package index.
function failingLine(stack: string, oathPath: string): number | undefined {
  const escaped = oathPath.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
  const m = new RegExp(`${escaped}:(\\d+):\\d+`).exec(stack)
  return m ? Number(m[1]) : undefined
}

// A thrown step error → the ExampleResult.failure payload. Shared by every
// producer (the vitest worker wrapper and the browser runner) so failures are
// byte-identical. Called only on the failure path, so it always returns a
// payload.
export function toFailure(
  error: unknown,
  oathPath: string,
  fallbackLine: number,
): NonNullable<ExampleResult['failure']> {
  const e = error as { message?: unknown; stack?: unknown }
  const stack = typeof e?.stack === 'string' ? e.stack : String(error)
  const message = e?.message != null ? String(e.message) : String(error)

  const cells: ReadonlyArray<CellFailure> | undefined = isCellMismatchError(error)
    ? error.cells
        .filter((c) => !c.ok)
        .map((c) => ({ from: c.span.startOffset, to: c.span.endOffset, actual: c.actual }))
    : undefined

  // executePlan attached the anchor when it caught the error, so this is the
  // failing step's span (or the first mismatched cell's). Absent only when the
  // error never passed through a step — then `line` is all a renderer gets.
  const anchor = readFailureAnchor(error)

  return {
    line: failingLine(stack, oathPath) ?? fallbackLine,
    message,
    stack,
    ...(cells && cells.length > 0 ? { cells } : {}),
    ...(anchor ? { anchor: { from: anchor.startOffset, to: anchor.endOffset } } : {}),
  }
}
